File: src/lib/email/threading.ts
const MESSAGE_ID_PATTERN = /<[^<>\s]+>/g;
const SUBJECT_PREFIX_PATTERN = /^\s*(?:re|fw|fwd)\s*:\s*/i;

export function extractMessageIds(value: string | undefined): string[] {
  return value?.match(MESSAGE_ID_PATTERN) ?? [];
}

export function getHeader(
  headers: Record<string, string> | undefined,
  name: string,
): string | undefined {
  if (!headers) {
    return undefined;
  }

  const target = name.toLowerCase();
  return Object.entries(headers).find(
    ([headerName]) => headerName.toLowerCase() === target,
  )?.[1];
}

export function normalizeSubject(subject: string): string {
  let normalized = subject.trim();
  while (SUBJECT_PREFIX_PATTERN.test(normalized)) {
    normalized = normalized.replace(SUBJECT_PREFIX_PATTERN, '');
  }
  return normalized || subject.trim();
}

export function createReplySubject(subject: string): string {
  return `Re: ${normalizeSubject(subject)}`;
}

export function buildReferences(
  references: string[],
  parentInternetMessageId: string,
): string[] {
  return [...new Set([...references, parentInternetMessageId])];
}

export function parseAddress(value: string): {
  address: string;
  name: string | null;
} {
  const match = value.trim().match(/^(.*?)\s*<([^<>]+)>$/);
  if (!match) {
    return { address: value.trim(), name: null };
  }

  const name = match[1].trim().replace(/^"|"$/g, '');
  return { address: match[2].trim(), name: name || null };
}

File: src/main/kotlin/com/castab/resend/email/Threading.kt
package com.castab.resend.email

private val messageId = Regex("<[^<>\\s]+>")
private val subjectPrefix = Regex("^\\s*(?:re|fw|fwd)\\s*:\\s*", RegexOption.IGNORE_CASE)

fun extractMessageIds(value: String?): List<String> = value?.let { messageId.findAll(it).map(MatchResult::value).toList() }.orEmpty()

fun normalizeSubject(value: String): String {
    var normalized = value.trim()
    while (subjectPrefix.containsMatchIn(normalized)) normalized = subjectPrefix.replaceFirst(normalized, "")
    return normalized.ifBlank { value.trim() }
}

fun replySubject(value: String) = "Re: ${normalizeSubject(value)}"
fun createReplySubject(value: String) = replySubject(value)
fun buildReferences(ancestry: List<String>, parent: String) = (ancestry + parent).distinct()

/** Case-insensitive header lookup over a raw header map (mirrors the Express `getHeader`). */
fun getHeader(headers: Map<String, String>?, name: String): String? {
    if (headers == null) return null
    val target = name.lowercase()
    return headers.entries.firstOrNull { it.key.lowercase() == target }?.value
}

data class Mailbox(val address: String, val name: String?)
fun parseAddress(value: String): Mailbox {
    val match = Regex("^(.*?)\\s*<([^<>]+)>$").matchEntire(value.trim())
        ?: return Mailbox(value.trim(), null)
    return Mailbox(match.groupValues[2].trim(), match.groupValues[1].trim().trim('"').ifBlank { null })
}

